### ISSUE TITLE
Behavior during training 

### ISSUE THREAD

**Issue by Kadakol** [tech-srl/code2vec#87]
Hi @urialon 

Thank you for your work with code2vec!

I was trying to understand better how code2vec works. However, I noticed some discrepancies which I have not been able to explain. Could you please help me understand?

During training using the java14m dataset, the training went fine till 14 epochs. However, a few minutes after the 14th epoch, the training suddenly ended. The target was 20 epochs. The only change from default configuration was the batch size (256) to avoid OOM. The F1-score after the 14th epoch was 0.4974 and on the test set was 0.4964. Also, I was using default framework (tensorflow) for points 1 to 3 below.

1. Why did the training end after a few batches after the 14th epoch? I see no errors in the logs, so the termination seems to be graceful. If it is because F1 score did not improve over certain number of epochs, I would have expected the evaluation to be conducted immediately after the 14th epoch and not some time later. Please let me know what I am missing here.

2. The model provided in the [repo](https://github.com/tech-srl/code2vec#downloading-a-trained-model-14-gb) gives me an F1 score of 0.588 (similar to what is reported in the paper). However, at no point when I was training the model till 14 epochs was I able to see such a high score. Could you please help me understand why I am not able to reproduce the same results? (Changing the batch size to 256 could not have had this effect, right?).

3. Only 6 models are being saved even though the config has a value of 10. Moreover, these are not the latest 6 models. Following is the pattern I am seeing, in terms of models saved after that many epochs:
    1 2 3 4 5 6
    1 3 4 5 6 7
    1 4 5 6 7 8
    .
    .
    .
    1 10 11 12 13 14
I did see the following warning:
```
# This warning was observed:
	WARNING:tensorflow:From /home/ubuntu/anaconda3/envs/tensorflow_gpu_env/lib/python3.6/site-packages/tensorflow_core/python/training/saver.py:963: remove_checkpoint (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
	Instructions for updating:
	Use standard file APIs to delete files with this prefix.
```
Perhaps I am missing something in my setup?

4. How do I evaluate the saved models using keras as the framework? The following does not seem to work.
`python3 code2vec.py --load models/java14m/saved_model__entire-model/ckpt-11 --test data/java14m/java14m.test.c2v --framework keras`
I get the error
`ValueError: Model dictionaries file is not found in model load dir. Expecting file models/java14m/saved_model__entire-model/dictionaries.bin.`
I see dictionaries.bin is in models/java14m/. So I copied it to models/java14m/saved_model__entire-model/. Then I get the following error.
`ValueError: There is no entire model to load at path models/java14m/saved_model__entire-model/ckpt-11__entire-model, and there is no model weights file to load at path models/java14m/saved_model__entire-model/ckpt-11__only-weights.`
Following are some of the files in the directory:
```
3.2G Jun 15 15:24 ckpt-11.data-00000-of-00002
943M Jun 15 15:24 ckpt-11.data-00001-of-00002
1.7K Jun 15 15:24 ckpt-11.index
968 Jun 15 15:24 checkpoint
```
I am stuck at this point. I could not find any documentation for this (maybe I overlooked it). Request you to help me understand how to proceed with this.

Thank you.

**Comment by urialon** [comment]
Hi @Kadakol ,
Thank you for your interest in code2vec!

Regarding point 1: Can you attach the training log that was printed to the console?

Regarding point 2: the smaller batch size might be related to this, but I'll check. I'll kick a new training run with batch size of 512.

Regarding point 3: only the last 5 best models are saved, to save disk space. This is defined here: [https://github.com/tech-srl/code2vec/blob/master/tensorflow_model.py#L57](https://github.com/tech-srl/code2vec/blob/master/tensorflow_model.py#L57)

Regarding point 4: please don't use the Keras version, I'm sorry. I think I will revert it.

**Comment by Kadakol** [comment]
Hi @urialon ,

Thank you so much for your quick reply!

> Regarding point 1: Can you attach the training log that was printed to the console?

Unfortunately I do not have the logs for the exact scenario I have reported. I do have the logs for a different run which also displays somewhat similar behavior. I have trimmed it down only to the significant parts in the attached file. As you can see, each epoch takes ~50 minutes. But you can see that the training terminated ~30 minutes after the 8th epoch. By the way, you can also see the warning that I mentioned about in point 3 above. [Code2Vec_Issue#87_Point1.txt](https://github.com/tech-srl/code2vec/files/4784296/Code2Vec_Issue.87_Point1.txt)

> Regarding point 2: the smaller batch size might be related to this, but I'll check. I'll kick a new training run with batch size of 512.

OK. Thank you. Please do let me know what you find out.

> Regarding point 3: only the last 5 best models are saved, to save disk space. This is defined here: https://github.com/tech-srl/code2vec/blob/master/tensorflow_model.py#L57

Thank you for pointing this out. I see that it is referring to the config MAX_TO_KEEP. I was under the impression that the last 10 models would be saved, as defined in the [README](https://github.com/tech-srl/code2vec#configmax_to_keep--10). Also, how come model number 1 never gets deleted? Only the next 5 models seem to be continuously rolled over. The model after the 1st iteration has the lowest F1 score among all the trained models, so I don't think it is being retained because it is one of the best.

> Regarding point 4: please don't use the Keras version, I'm sorry. I think I will revert it.

Got it. Thank you.

**Comment by urialon** [comment]
Hi @Kadakol ,
I kicked a training run with the default settings for 8 epochs, and so far it had reached F1=57.89 on the validation set, which is very close already to the best results.
So I suspect that the small batch size is the problem in your case. 
Maybe reducing the learning rate by 0.5-0.25 to keep it proportional to the batch size will help.

**Comment by Kadakol** [comment]
Hi @urialon ,

> I kicked a training run with the default settings for 8 epochs, and so far it had reached F1=57.89 on the validation set, which is very close already to the best results.

Thank you so much for trying this out!

> So I suspect that the small batch size is the problem in your case.
> Maybe reducing the learning rate by 0.5-0.25 to keep it proportional to the batch size will help.

Please correct me if I'm wrong here. Your suggestion is to change the line https://github.com/tech-srl/code2vec/blob/master/tensorflow_model.py#L232. The default learning rate is 0.001. Maybe changing this to 0.0005 or 0.00025 should help. So the change should be something like

`optimizer = tf.compat.v1.train.AdamOptimizer(learning_rate=0.00025).minimize(loss)`

Is my understanding correct?

**Comment by urialon** [comment]
Exactly, but it might require further tuning. 

Regarding point 1:
I looked again at the code - I do not see a clear bug in the computation of when to stop. But just in case, you can set the number of epochs to 9999 or something (and stop manually by killing the process), such that it won't stop unexpectedly. Let me know if it still stops unexpectedly even when `NUM_TRAIN_EPOCHS=9999`.

Regarding point 3 - the saving of files:
Yes, the 1st checkpoint is saved and never deleted, while in the following iterations only the last 5 are kept. This is indeed a minor bug.

**Comment by Kadakol** [comment]
> Exactly, but it might require further tuning.

Let me try this out.

> Regarding point 1:
> I looked again at the code - I do not see a clear bug in the computation of when to stop. But just in case, you can set the number of epochs to 9999 or something (and stop manually by killing the process), such that it won't stop unexpectedly. Let me know if it still stops unexpectedly even when `NUM_TRAIN_EPOCHS=9999`.
> 
> Regarding point 3 - the saving of files:
> Yes, the 1st checkpoint is saved and never deleted, while in the following iterations only the last 5 are kept. This is indeed a minor bug.

OK. Thank you!

**Comment by Kadakol** [comment]
Hi @urialon ,

> > Exactly, but it might require further tuning.
> 
> Let me try this out.

I'm finding it difficult to do this, especially because the training process is crashing in between. I've attached the log file -[java14m-lr0.00025-b256.log.txt](https://github.com/tech-srl/code2vec/files/4796068/java14m-lr0.00025-b256.log.txt)

If I use a batch size of 1024, the crash would happen in a matter of seconds, before any training can start.

If I use a batch size of 256, training will take ~1h 20m per epoch. 10 minutes after the 2nd epoch, this OOM happened.

Any idea why I would get an OOM in between epochs? I am using an RTX 2080 Ti with 11GB of memory?

**Comment by urialon** [comment]
Sorry, I don't know. I originally trained the model on an Nvidia K80, and verified a few days ago on a V100.

**Comment by Kadakol** [comment]
Hi @urialon ,

A quick update. I finally managed to try this out (https://github.com/tech-srl/code2vec/issues/87#issuecomment-645125074). I set the batch size to 256, the learning rate to 0.25e-3 and the number of epochs to 80. Training terminated after 57 epochs. Following is the trend I see w.r.t. validation accuracy.

After 1 epoch - F1 score is 0.4124
After 8 epochs - F1 score is 0.50 (reached 0.5 for the first time)
After 20 epochs - F1 score hits 0.5116
After 51 epochs - F1 score hits a peak of 0.5122

This definitely beats my previous run with batch size 256, learning rate 1e-3 (default) and 20 epochs (default) where the max F1 score was 0.48. But it still does not reach the values reported in the paper.

As you had mentioned, I guess it requires a bit more fine-tuning. 

Thank you for pointing me in this direction!
